### ISSUE TITLE
Non-mutually exclusive outcomes

### ISSUE THREAD

**Issue by BrianMiner** [dnouri/nolearn#80]
Is it possible to have non mutually exclusive outcomes in nolearn / lasagne? Specifically where there is a vector of outcomes for each training observation? Something along these lines? What would be a suitable loss function?

import numpy as np

X=np.random.random((50, 10))
y=np.random.randint(2, size=(50,3))
y=np.cast['int32'](y)

num_classes = 3
num_features = X.shape[1]

layers0 = [('input', InputLayer),
           ('dense0', DenseLayer),
           ('output', DenseLayer)]

net0 = NeuralNet(layers=layers0,
                 
                 input_shape=(None, num_features),  
                 dense0_num_units=3, 
                 
                 output_num_units=num_classes,
                 output_nonlinearity=sigmoid,
                 
                 objective_loss_function =categorical_crossentropy,
                 
                 update=nesterov_momentum,
                 update_learning_rate=0.5,
                 update_momentum=0.9,
                 
                 eval_size=0,
                 verbose=1,
                 max_epochs=50)

net0.fit(X, y)

**Comment by benanne** [comment]
Try using `binary_crossentropy` instead, that makes more sense :) This problem setting is typically called multi-label classification by the way.

**Comment by BrianMiner** [comment]
Can this be done currently - can a vector of the target variable, as shown above, i.e. y=np.random.randint(2, size=(50,3))
y=np.cast'int32'
?

**Comment by swarbrickjones** [comment]
I've just done this, it seems to work OK.

N.B. you need to leave the regression flag = True, otherwise the stratified crossvalidation nolearn uses will throw an error.  Also remember to put a nonlinearity on the output to fix it to (0,1) , NOT softmax, I used sigmoid.

Here's the net I used - 

```python
ae = NeuralNet(
    layers=[
        ('input', layers.InputLayer),
        ('conv10', layers.Conv2DLayer),
        #('conv11', layers.Conv2DLayer),
        ('pool1', layers.MaxPool2DLayer),
        ('conv20', layers.Conv2DLayer),
        ('conv21', layers.Conv2DLayer),
        ('pool2', layers.MaxPool2DLayer),
        ('hidden1', layers.DenseLayer),
        ('hidden2', layers.DenseLayer),
        #('hidden3', layers.DenseLayer),
        ('output', layers.DenseLayer),
        ],
    input_shape=(None, 3, 80, 60),
    conv10_num_filters=16, 
    conv10_filter_size=(3, 3), 
    conv10_border_mode="valid",
    #conv11_num_filters=32, 
    #conv11_filter_size=(3, 3), 
    #conv11_border_mode="valid",
    pool1_ds=(2, 2),
    conv20_num_filters=64, 
    conv20_filter_size=(5, 5), 
    conv20_border_mode="valid",
    pool2_ds=(2, 2),
    conv21_num_filters=64, 
    conv21_filter_size=(5, 5), 
    conv21_border_mode="valid",
    hidden1_num_units = 1048,
    hidden2_num_units = 1048,
    #hidden3_num_units=512,
    output_num_units = y_train.shape[1],
    output_nonlinearity = lasagne.nonlinearities.sigmoid,
    update_learning_rate=0.01,
    update_momentum=0.975,
    regression = True,
    batch_iterator_train=FlipBatchIterator(batch_size=128),
    objective_loss_function=lasagne.objectives.binary_crossentropy,
    max_epochs= 100,
    verbose=1,
    )
```

**Comment by mikesj-public** [comment]
I've just done this, it seems to work OK.

N.B. you need to leave the regression flag = True, otherwise the stratified crossvalidation nolearn uses will throw an error.  Also remember to put a nonlinearity on the output to fix it to (0,1) , NOT softmax, I used sigmoid.

Here's the net I used - 

```python
ae = NeuralNet(
    layers=[
        ('input', layers.InputLayer),
        ('conv10', layers.Conv2DLayer),
        #('conv11', layers.Conv2DLayer),
        ('pool1', layers.MaxPool2DLayer),
        ('conv20', layers.Conv2DLayer),
        ('conv21', layers.Conv2DLayer),
        ('pool2', layers.MaxPool2DLayer),
        ('hidden1', layers.DenseLayer),
        ('hidden2', layers.DenseLayer),
        #('hidden3', layers.DenseLayer),
        ('output', layers.DenseLayer),
        ],
    input_shape=(None, 3, 80, 60),
    conv10_num_filters=16, 
    conv10_filter_size=(3, 3), 
    conv10_border_mode="valid",
    #conv11_num_filters=32, 
    #conv11_filter_size=(3, 3), 
    #conv11_border_mode="valid",
    pool1_ds=(2, 2),
    conv20_num_filters=64, 
    conv20_filter_size=(5, 5), 
    conv20_border_mode="valid",
    pool2_ds=(2, 2),
    conv21_num_filters=64, 
    conv21_filter_size=(5, 5), 
    conv21_border_mode="valid",
    hidden1_num_units = 1048,
    hidden2_num_units = 1048,
    #hidden3_num_units=512,
    output_num_units = y_train.shape[1],
    output_nonlinearity = lasagne.nonlinearities.sigmoid,
    update_learning_rate=0.01,
    update_momentum=0.975,
    regression = True,
    batch_iterator_train=FlipBatchIterator(batch_size=128),
    objective_loss_function=lasagne.objectives.binary_crossentropy,
    max_epochs= 100,
    verbose=1,
    )
```

**Comment by BrianMiner** [comment]
@mikesj-public Thank you for the information. So your objective function is binary cross entropy and and sigmoid output activation. What is the nature of the output you are predicting?  output_num_units = y_train.shape[1] relates to a vector output?

**Comment by mikesj-public** [comment]
Yeah, y_train is a list of binary vectors, which correspond to one or more descriptive words about the image.

**Comment by BrianMiner** [comment]
@mikesj-public  many thanks!

**Comment by BrianMiner** [comment]
@mikesj-public I was curious about this again - trying to formulate the data correctly. Is y_train actually a list (since a list does not have a shape property I dont believe).

**Comment by mikesj-public** [comment]
Hey Brian.  y_train is a 2d numpy array, a list of lists.  Say I had N samples.  Each sample corresponds to a list of 0s and 1s.  In my particular case I was looking at two different features of a picture at once as my output - the colour and the shape.  So I made a one-hot encoded array of the colours (an N x 7 array, say), and a one-hot encoded array for the shapes (an N x 20 array, say), and then concatenated them together to give an N x 27 array, where each row has exactly two 1s.  Hope that helps.

**Comment by BrianMiner** [comment]
Mike, is the number of output_num_units = 27?
